# Project Title
Comparison of Logistic Regression and SVM Method

## Members

1. First member: name and e-mail adress
Zeynep Göger and gogerz@itu.edu.tr
2. Second member: name and e-mail adress

## Description of the project
In this project,  we compare two methods that are logistic regression method and SVM method 
by using 1984 United Stated Congressional Voting Records in R language. We analyse that which
method is better to predict the congressman belongs to which party according to first column
of the data to the rest of them.


### The methods to be used
There are two different methods that are logistic regression method and SVM method.
Logistic regression is a type of a prediction method that is used if the data has binary 
dependent variables and this method determines the relationship between the set of dependent
variables and the independent set.
    
Support Vector Machines method is another prediction method that is used if the data is
binary too. However, in this method two kinds of accumulation of dependent variables
are determined and divided down the middle of the distance that is called as margin.    

Explain.

### The data

The data is 1984 United Stated Congressional Voting Records.This data consists of 435 congressmen that are 267 democrats and 168 republicans.
In this data has binary dependent variables such as yes or no and two class of congressmen
that are republican and democrat. There are 16 numbers of attributes that are categorical.
    
## Code

As proof of work, you must run this notebook.  Upload an HTML output of this notebook on your github account.

In [5]:
X<- read.csv("https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data", header=FALSE)

In [6]:
head(X)

V1         V2 V3 V4 V5 V6 V7 V8 V9 V10 V11 V12 V13 V14 V15 V16 V17
1 republican n  y  n  y  y  y  n  n  n   y   ?   y   y   y   n   y  
2 republican n  y  n  y  y  y  n  n  n   n   n   y   y   y   n   ?  
3 democrat   ?  y  y  ?  y  y  n  n  n   n   y   n   y   y   n   n  
4 democrat   n  y  y  n  ?  y  n  n  n   n   y   n   y   n   n   y  
5 democrat   y  y  y  n  y  y  n  n  n   n   y   ?   y   y   y   y  
6 democrat   n  y  y  n  y  y  n  n  n   n   n   n   y   y   y   y

## Logistic Regression
Logistic regression is a type of a prediction method that is used if the data has binary 
dependent variables and this method determines the relationship between the set of dependent
variables and the independent set. Odds is the ratio of the probability of occurrence with probability of event that does not occur.

$logit_p = log(odds) = log(p/(1-p)) =  b_0 + b_1x_1 +...+  b_k x_k$

where $b_k's$ are constant coefficients that determines the changes of logistic regression model when $x_i$ is added.

Initially, we divide the data to constitute our model and we divide 30% of the data randomly.


To do this we use this code 


In [13]:
sample(1:nrow(X), 0.3*nrow(X))
test <-sample(1:nrow(X), 0.3*nrow(X))

[1]  65 325  16 241  68 146 383 348 390 376 259 264 361 111 326 238 147  33
 [19]  18 251  11 342 367  78 385 228 271 375 202  58 306 296 175 116 207  53
 [37]  83 171 234 298 275   2 224 332 428  89  80 190 304 416 211 313 268  28
 [55]  31 112 210 316  25 263 315 396   3 194 347 158 430  24 291 401 102 321
 [73]   1 394 220 290 434 219 349 419 249 427  77 425 392  74 329  46 206 122
 [91] 380 139 172  61 173 410  45 331 244 222 395 411 283 328 311 292 322 320
[109] 351 247 343 365 384 284 308  63  32 144 235 130 309 360 223  94   9  48
[127] 254 270 149 337

In [14]:
length(test)


[1] 130

130 members of this data is selected randomly by using this code in R but these numbers are not useful for testing and forming our model. Therefore, we convert them to a matrix by using the code below.

In [15]:
X[test,]

V1         V2  V3  V4  V5  V6  V7  V8  V9  V10 V11 V12 V13 V14 V15 V16 V17
68  republican n   y   n   y   y   y   n   n   n   y   n   y   y   y   n   n  
14  democrat   y   y   y   n   n   y   y   y   ?   y   y   ?   n   n   y   ?  
216 democrat   n   y   y   y   y   y   n   n   n   y   y   y   y   y   y   ?  
432 democrat   n   n   y   n   n   n   y   y   y   y   n   n   n   n   n   y  
244 democrat   n   y   y   n   n   n   y   y   ?   y   n   n   y   n   y   y  
289 democrat   y   y   y   n   y   y   n   y   n   y   y   n   y   y   y   y  
158 democrat   n   y   y   n   n   y   n   y   y   n   y   n   ?   ?   ?   ?  
92  democrat   y   n   y   n   n   n   y   y   y   y   y   n   n   n   y   y  
159 republican n   y   n   y   y   y   n   n   n   y   n   y   y   y   n   n  
206 democrat   n   n   y   n   n   y   y   y   y   y   n   y   n   n   n   y  
362 democrat   y   n   y   n   y   y   y   n   y   y   n   n   y   y   n   ?  
53  democrat   y   y   y   n   n   n   y   y   y   n   n   n   n   n   ?   ?  
250 democrat   y   n   y   n   n   n   y   y   ?   n   y   n   n   n   y   y  
183 democrat   y   ?   y   n   n   n   y   y   y   n   n   n   n   n   y   y  
397 democrat   y   y   y   n   y   y   n   y   n   n   y   n   y   n   y   y  
288 democrat   y   y   y   ?   n   y   y   y   y   n   y   n   y   n   ?   y  
305 republican n   y   n   y   y   y   n   n   n   y   n   y   y   y   n   ?  
57  republican n   y   n   y   y   y   n   n   n   y   y   y   y   y   n   y  
368 democrat   y   y   y   n   y   y   y   y   n   y   y   n   n   n   y   ?  
99  democrat   y   y   y   n   n   y   y   y   y   y   n   n   n   n   n   y  
359 democrat   y   y   y   n   n   y   y   y   y   y   y   y   y   y   n   ?  
364 republican y   y   n   y   y   y   n   n   n   y   y   n   y   n   n   n  
193 democrat   n   n   n   n   n   y   y   y   y   n   y   n   n   y   y   y  
113 democrat   n   ?   y   n   n   y   y   y   n   y   n   n   n   n   y   ?  
160 democrat   n   y   y   n   ?   y   y   y   y   y   y   n   n   ?   n   ?  
365 republican y   y   n   y   y   y   n   n   n   n   y   n   y   y   n   y  
291 democrat   y   n   y   n   ?   y   ?   y   y   y   n   n   y   y   n   y  
30  democrat   y   y   y   n   n   n   y   y   y   n   y   n   n   n   y   y  
55  democrat   y   y   y   n   n   y   ?   y   n   n   y   y   n   y   n   ?  
275 republican y   n   n   y   y   n   y   n   n   y   n   n   n   y   y   y  
... ...        ... ... ... ... ... ... ... ... ... ... ... ... ... ... ... ...
270 democrat   y   n   y   n   n   n   y   y   y   n   n   n   n   n   n   y  
234 republican n   y   n   y   y   y   n   n   n   y   n   y   y   y   n   y  
260 democrat   y   n   y   n   n   n   y   y   y   n   n   n   n   n   n   y  
122 republican n   ?   n   y   y   y   n   n   n   y   n   y   y   y   n   y  
346 republican n   n   n   y   y   y   y   n   n   y   n   y   n   y   y   y  
214 democrat   n   y   y   n   n   y   n   y   y   n   y   n   y   n   y   y  
170 democrat   y   n   y   n   n   n   y   y   y   y   y   n   n   y   n   y  
59  republican n   y   n   y   y   y   n   n   n   y   n   y   y   y   n   y  
126 republican n   ?   n   y   y   y   n   n   n   y   n   y   y   y   n   y  
174 democrat   n   n   n   n   y   y   n   n   n   y   y   y   y   y   n   y  
134 republican n   n   n   y   y   y   n   n   n   y   n   y   y   y   n   y  
130 democrat   ?   ?   y   n   n   n   y   y   ?   n   ?   ?   ?   ?   ?   ?  
308 democrat   y   n   y   n   n   y   y   y   y   n   n   n   n   y   n   ?  
178 democrat   n   n   y   n   n   n   y   y   y   y   y   ?   n   n   y   y  
222 democrat   n   y   y   n   y   y   n   n   n   n   y   n   n   n   y   ?  
180 democrat   y   n   y   n   n   n   y   y   y   y   n   n   n   n   y   y  
140 democrat   y   n   y   n   n   y   y   y   y   n   n   n   n   n   y   y  
196 republican n   n   n   y   y   y   n   n   n   n   n   y   y   y   n   ?  
212 republican n   y   n   y   y   y   n   n   n   y  

In [22]:
nrow(X[test,])
modeldata<-X[test,]

[1] 130

As shown above, the numbers of model are brought together by converting to the X matrix and it is assigned to modeldata variable by using assignment operator of R.
To convert the rest of them, we use this code written in the below.


In [18]:
X[-test,]

V1         V2  V3  V4  V5  V6  V7  V8  V9  V10 V11 V12 V13 V14 V15 V16 V17
1   republican n   y   n   y   y   y   n   n   n   y   ?   y   y   y   n   y  
2   republican n   y   n   y   y   y   n   n   n   n   n   y   y   y   n   ?  
3   democrat   ?   y   y   ?   y   y   n   n   n   n   y   n   y   y   n   n  
5   democrat   y   y   y   n   y   y   n   n   n   n   y   ?   y   y   y   y  
7   democrat   n   y   n   y   y   y   n   n   n   n   n   n   ?   y   y   y  
8   republican n   y   n   y   y   y   n   n   n   n   n   n   y   y   ?   y  
9   republican n   y   n   y   y   y   n   n   n   n   n   y   y   y   n   y  
12  republican n   y   n   y   y   y   n   n   n   n   y   ?   y   y   ?   ?  
13  democrat   n   y   y   n   n   n   y   y   y   n   n   n   y   n   ?   ?  
15  republican n   y   n   y   y   y   n   n   n   n   n   y   ?   ?   n   ?  
16  republican n   y   n   y   y   y   n   n   n   y   n   y   y   ?   n   ?  
17  democrat   y   n   y   n   n   y   n   y   ?   y   y   y   ?   n   n   y  
19  republican n   y   n   y   y   y   n   n   n   n   n   ?   y   y   n   n  
20  democrat   y   y   y   n   n   n   y   y   y   n   y   n   n   n   y   y  
21  democrat   y   y   y   n   n   ?   y   y   n   n   y   n   n   n   y   y  
23  democrat   y   ?   y   n   n   n   y   y   y   n   n   ?   n   n   y   y  
24  democrat   y   y   y   n   n   n   y   y   y   n   n   n   n   n   y   y  
25  democrat   y   n   y   n   n   n   y   y   y   n   n   n   n   n   y   ?  
27  democrat   y   n   y   n   n   n   y   y   y   n   y   n   n   n   y   y  
31  republican n   y   n   y   y   y   n   n   n   n   n   y   y   y   n   n  
32  democrat   y   y   y   n   n   n   y   y   y   n   y   n   n   n   y   ?  
33  democrat   y   y   y   n   n   n   y   y   y   y   n   n   y   n   y   y  
34  republican n   y   n   y   y   y   n   n   n   n   n   y   y   y   n   y  
35  democrat   y   y   y   n   n   n   y   y   y   n   n   n   n   n   y   y  
36  republican n   y   n   y   y   y   n   n   n   n   n   y   y   y   n   n  
37  republican y   ?   n   y   y   y   n   n   n   y   n   y   ?   y   n   y  
38  republican y   y   n   y   y   y   n   n   n   n   n   n   y   y   n   y  
39  republican n   y   n   y   y   y   n   n   n   y   n   y   y   y   n   n  
41  democrat   y   y   y   n   n   n   y   y   y   n   ?   n   n   n   n   ?  
42  democrat   y   y   y   n   n   n   y   y   y   n   n   n   n   n   y   ?  
... ...        ... ... ... ... ... ... ... ... ... ... ... ... ... ... ... ...
400 republican n   y   n   y   ?   y   n   n   n   y   n   y   y   y   n   n  
401 republican n   y   n   y   y   y   n   ?   n   n   ?   ?   ?   y   n   ?  
402 republican n   y   n   y   y   y   n   n   n   y   y   y   y   y   n   n  
403 republican ?   n   y   y   n   y   y   y   y   y   n   y   n   y   n   y  
404 republican n   y   n   y   y   y   n   n   n   y   n   y   ?   y   n   n  
405 republican y   y   n   y   y   y   n   n   n   y   n   y   y   y   n   y  
406 republican n   n   n   y   y   y   n   n   n   n   n   y   y   y   n   y  
408 democrat   n   n   n   y   y   y   n   n   n   n   y   y   y   y   n   n  
409 democrat   y   n   y   n   n   y   y   y   y   n   n   y   ?   y   y   y  
410 republican n   n   n   y   y   y   n   n   n   n   n   y   y   y   n   n  
411 republican n   n   n   y   y   y   n   n   n   n   y   y   y   y   n   y  
413 republican n   n   n   y   y   y   n   n   n   y   n   y   y   y   n   y  
415 democrat   y   y   y   n   n   n   y   y   y   n   n   n   n   n   n   y  
416 democrat   n   y   y   n   n   y   y   y   ?   y   n   n   n   n   n   y  
417 republican y   y   n   y   y   y   n   n   n   y   n   n   y   y   n   y  
418 democrat   y   y   y   n   n   n   y   y   y   y   y   n   y   n   n   y  
420 democrat   y   y   y   n   n   n   y   y   y   n   n   n   n   n   n   y  
422 democrat   n   y   y   n   y   y   y   y   n   n   y   n   y   n   y   y  
423 democrat   n   n   y   n   n   y   y   y   y   n  

We use the rest of the model data to test our model;consequently, it is assigned to testdata 
variable. 

In [33]:
testdata<-X[-test,]

Logistic regression model is formed by using glm function in R language. The meaning of glm is fitting  generalized linear models.

In [34]:
logitmodel<- glm(V1 ~ . , family=binomial(link="logit"), data= modeldata)

Warning message:
"glm.fit: algorithm did not converge"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"

After the model is generated, we use summary code to get a short information about the model 
that we generate.

In [35]:
summary(logitmodel)


Call:
glm(formula = V1 ~ ., family = binomial(link = "logit"), data = modeldata)

Deviance Residuals: 
       Min          1Q      Median          3Q         Max  
-2.719e-05  -2.110e-08  -2.110e-08   2.110e-08   2.836e-05  

Coefficients:
              Estimate Std. Error z value Pr(>|z|)
(Intercept)  7.974e+01  4.036e+05       0        1
V2n          2.065e+01  8.762e+05       0        1
V2y          3.208e+01  7.510e+05       0        1
V3n         -1.950e+00  1.800e+05       0        1
V3y         -3.051e+01  1.823e+05       0        1
V4n         -4.979e+01  4.509e+05       0        1
V4y         -3.456e+01  4.482e+05       0        1
V5n          5.807e+00  7.610e+05       0        1
V5y          1.139e+02  9.204e+05       0        1
V6n          5.240e+00  4.023e+05       0        1
V6y         -2.835e-01  4.430e+05       0        1
V7y          1.848e+01  1.673e+05       0        1
V8n          1.532e+01  3.684e+05       0        1
V8y          1.928e+01  3.488e+05       0    

After the model is formed, we use predict function to estimate the model for testdata variable 
which is the rest of the modeldata.

In [43]:

predict(logitmodel,testdata, type = c("link", "response", "terms"))


ERROR: Error in model.frame.default(Terms, newdata, na.action = na.action, xlev = object$xlevels): factor V7 has new levels ?
